### ISSUE TITLE
[Bug]: CodeLite loses its compiler and complains "Some of the compilers referred by the workspace no longer exist. Define each missing compiler by cloning an existing compiler."

### ISSUE THREAD

**Issue by vpp612** [eranif/codelite#3524]
### What happened?

I had reorganized the folders containing C++ training material so that each project resides under a workspace folder named by the lesson name. I noticed some errors in the training material, e.g. .mk and .project files were named differently from the project name and some project names had spelling mistakes. After organizing, some of the projects were not possible to compile and produced the error above. Since then I have tried to run the configuration wizard several times and configured separately some projects, but to no avail.
Additionally, the active project is not always the one being built, but the previous or some older. This is very odd.
My compiler setting is "GCC".

### Version

14.0.0

### Operating system

Linux

### Steps to reproduce

```text
Running CodeLite which has been configured earlier during a programming course and has earlier worked properly. 

Selecting Build Project, Compile Current File, Compile Current File's Project, Clean Project in Build drop-down. Also some selections available with mouse right-button on the file project produce the same error.
```

### Relevant log output

```shell
I cannot find any relevant log output.
```

**Comment by Jarod42** [comment]
Providing problematic projects might help...
We don't know what changes you did.

(I tend to use project generators (such as CMake/premake) to generates project.)

**Comment by vpp612** [comment]
Hello, I would like the IDE (here CodeLite) to explain better why the compiler is missing, because it has been configured. How should it be configured so that it is valid?

Cheers,
Veli-Pekka

**Comment by Jarod42** [comment]
As I understand, you had a working project, did some changes, then it no longer works.
I would say your changes are invalid...
I don't know how to help with so few information...

**Comment by eranif** [comment]
What happened is this:

- Initially, you had a compiler that CodeLite named it "Compiler A"
- You did some upgrade and some point you let CodeLite re-scan OR you opened the same workspace on a different machines that does not have the compiler named "Compiler A", instead it has "Compiler B"
- You tried to build your project, so CodeLite searched its configuration for "Compiler A". It could not be found, hence the message you are getting.

How to fix this?

3 ways:

1. The dialog that CodeLite shows is basically saying: "Hey, I don't have 'Compiler A', but I do have 'Compiler B', do you want me to create a clone of 'Compiler A' and name it 'Compiler B'?"
2. Go to "settings -> build settings" locate the best suited compiler from the list and re-name to match the compiler name that CodeLite is searching
3. Open your project settings and under the "General -> Compiler" select an existing compiler from the list.

Option 2 is the way I would go

**Comment by vpp612** [comment]
Thank you for your replies!
Anyway, the compiler is the same, but it may have been updated to a newer version. Could this cause such an issue? If not, then probably the new run of configuration wizard. I need to study the suggested fixes on Monday.

**Comment by eranif** [comment]
A quick search from the command line:

Open a terminal, `cd` to the workspace folder and type:

```bash
find . -name "*.project"|xargs grep -w --color CompilerType
```

This will produce an output similar to this:

```
$ find . -name "*.project"|xargs grep -w --color CompilerType
    <Configuration Name="Debug" CompilerType="MinGW 64bit ( MSYS2 64bit )" DebuggerType="GNU gdb debugger" Type="Executable" BuildCmpWithGlobalSettings="append" BuildLnkWithGlobalSettings="append" BuildResWithGlobalSettings="append">
    <Configuration Name="Release" CompilerType="MinGW 64bit ( MSYS2 64bit )" DebuggerType="GNU gdb debugger" Type="Executable" BuildCmpWithGlobalSettings="append" BuildLnkWithGlobalSettings="append" BuildResWithGlobalSettings="append">
M
```

Check the compiler names and make sure you have them all defined under: `settings -> build settings`

**Comment by vpp612** [comment]
Thanks a lot Eran! These were good hints!

There are a lot of projects in the workspace folder, so the output of the command to run is long. There are two kinds of output: in the other, the compiler is "GCC" and in the other "MinGW". I thing this already wrong, although I have tried to fix it:

$ find . -name "*.project" | xargs grep -w --color CompilerType
./STL_Lambdas/STL_Lambdas.project:    <Configuration Name="Debug" CompilerType="GCC" DebuggerType="GNU gdb debugger" Type="Executable" BuildCmpWithGlobalSettings="append" BuildLnkWithGlobalSettings="append" BuildResWithGlobalSettings="append">


./Section5/BasicIO/BasicIO.project:    <Configuration Name="Debug" CompilerType="MinGW ( MinGW )" DebuggerType="GNU gdb debugger" Type="Executable" BuildCmpWithGlobalSettings="append" BuildLnkWithGlobalSettings="append" BuildResWithGlobalSettings="append">

Trying compilation in project "STL_Lambdas" (not easy because first CodeLite tries to compile the previous project, although "STL_Lambdas" is already selected):
Popup header: Could not find selected compiler...

Then trying to compile "BasicIO" (although having a wrong setting):
Again, not compiling "BasicIO", but something else. Then activating an already active project:
Popup header: could not find selected compiler...

Then I tried just another project and now it again compiles something else, this time project "STL_Lambdas" and the output is in xterm window:

/home/vpp/.codelite/tmp/vpp/codelite-exec.sh: line 3: /b/home/vpp/Workspaces/Workspace2/build-Debug/bin/STL_Lambdas: No such file or directory
Hit any key to continue...

I may have removed the build-Debug folder in some earlier fixing attempt. I thought it was a temporary folder.

Settings -> build settings:

![image](https://github.com/user-attachments/assets/3ac54551-e2a4-4c4d-9a5d-05ed7f54d752)

Typical Linux compiler settings.

I am hoping you already see what is wrong in my settings. I just don't know how to fix this.

**Comment by vpp612** [comment]
Sorry for my spelling mistakes. It seems the find command outputs were lost because the lines were so long. Probably I need some guidance relating to those. But, just to know that in some of them the compiler was MinGW and in some other GCC should be enough.

**Comment by vpp612** [comment]
https://github.com/user-attachments/assets/ed0b0cc7-5bf2-4771-b1f4-25b92e5a1dd7

Trying to force "GCC", but cannot make it.

**Comment by vpp612** [comment]
Now I have a clue when the problems started. I had also downloaded the course material with the sample code files from the training centre. I missed some training sessions, so I copied the workspace folders of the training centre to my workspace and then added them as existing projects to my workspace. Those projects might well have had different compiler and other settings. And additionally that workspace was in Windows and my workspace in Linux.

**Comment by vpp612** [comment]
Maybe I need to remove all projects and empty the workspace and start adding projects again.


https://github.com/user-attachments/assets/978f86c9-d32d-4dcf-9ec5-712d3585e0ce

And use only the source files from downloaded projects.

**Comment by vpp612** [comment]
Ok, I found a simple solution to my issue.

In that popup that starts with "Could not find selected compiler...", I just selected "MinGW" to the slot on left and "GCC" to the slot on right. This is the first time the OK button was enabled. And this setting now works as a charm. I have not yet compiled every project, but so far it goes fine.

I also tested that shell script checking the CompilerType in project files. There are still both "GCC" and "MinGW" in the results. Maybe that it not so wrong after all.

Then I checked Settings > Build Settings and now there is a new line on Compilers tab: "MinGW (MinGW)" having the same contents as "GCC".

I really couldn't have guessed that I need to rename my compiler that way. Nevertheless, it fixed my issue.
